sycl/source/detail/usm/usm_impl.cpp

```diff
@@ -21,75 +23,99 @@ namespace usm {
 
 void *alignedAlloc(size_t Alignment, size_t Size, const context &Ctxt,
                    alloc Kind) {
-  std::shared_ptr<context_impl> CtxImpl = detail::getSyclObjImpl(Ctxt);
-  std::shared_ptr<USMDispatcher> Dispatch = CtxImpl->getUSMDispatch();
-  pi_context C = CtxImpl->getHandleRef();
-  pi_result Error;
   void *RetVal = nullptr;
-
-  switch (Kind) {
-  case alloc::host: {
-    RetVal = Dispatch->hostMemAlloc(C, nullptr, Size, Alignment, &Error);
-    break;
-  }
-  case alloc::device:
-  case alloc::shared:
-  case alloc::unknown: {
-    RetVal = nullptr;
-    Error = PI_INVALID_VALUE;
-    break;
-  }
+  if (Ctxt.is_host()) {
+    if (Alignment) {
+      RetVal = ::aligned_alloc(Alignment, Size);
+    }
+    else {
+      RetVal = ::malloc(Size);
+    }
+  } else {
+    std::shared_ptr<context_impl> CtxImpl = detail::getSyclObjImpl(Ctxt);
+    std::shared_ptr<USMDispatcher> Dispatch = CtxImpl->getUSMDispatch();
+    pi_context C = CtxImpl->getHandleRef();
+    pi_result Error;
+
+    switch (Kind) {
+    case alloc::host: {
+      RetVal = Dispatch->hostMemAlloc(C, nullptr, Size, Alignment, &Error);
+      break;
+    }
+    case alloc::device:
+    case alloc::shared:
+    case alloc::unknown: {
+      RetVal = nullptr;
+      Error = PI_INVALID_VALUE;
+      break;
+    }
+    }
+
+    // Error is for debugging purposes.
+    // The spec wants a nullptr returned, not an exception.
+    if (Error != PI_SUCCESS)
+      return nullptr;
   }
-
-  // Error is for debugging purposes.
-  // The spec wants a nullptr returned, not an exception.
-  if (Error != PI_SUCCESS) return nullptr;
-
   return RetVal;
 }
 
 void *alignedAlloc(size_t Alignment, size_t Size, const context &Ctxt,
                    const device &Dev, alloc Kind) {
-  std::shared_ptr<context_impl> CtxImpl = detail::getSyclObjImpl(Ctxt);
-  std::shared_ptr<USMDispatcher> Dispatch = CtxImpl->getUSMDispatch();
-  pi_context C = CtxImpl->getHandleRef();
-  pi_result Error;
-  pi_device Id;
   void *RetVal = nullptr;
-
-  switch (Kind) {
-  case alloc::device: {
-    Id = detail::getSyclObjImpl(Dev)->getHandleRef();
-    RetVal = Dispatch->deviceMemAlloc(C, Id, nullptr, Size, Alignment, &Error);
-    break;
+  if (Ctxt.is_host()) {
+    if (Alignment) {
+      RetVal = ::aligned_alloc(Alignment, Size);
```

Comment:
Can you reuse "our" aligned allocator here?

Reply:
Yes

Reply:
aligned_allocator supports custom alignment(it has function setAlignment), you can add constructor which takes alignment as well.

Reply:
Ok.